File: model/order.go
package model

import (
	"time"
)

type OrderStatus string

const (
	KeyOrder              = "order"
	OrderStatusPending    = "pending"
	OrderStatusDelivering = "delivering"
	OrderStatusFulfilled  = "fulfilled"
	OrderStatusRejected   = "rejected"
)

type Order struct {
	Id        string     `bson:"_id,omitempty"`
	QueueId   string     `bson:"queueId"`
	CreatedAt *time.Time `bson:"createdAt"`
	UpdatedAt *time.Time `bson:"updatedAt"`
	Status    OrderStatus
	Eta       *time.Time
	TotalSum  float64  `bson:"totalSum"`
	Items     []string // product ids
}

type OrderItem struct {
	Id         string `bson:"_id,omitempty"`
	Attributes []*KeyValue
}

func (o *Order) CalcTotalSum(resolve func(pid string) (*Product, error)) (float64, error) {
	var sum float64

	for _, it := range o.Items {
		p, err := resolve(it)
		if err != nil {
			return 0, err
		}

		sum += p.Price
	}

	return sum, nil
}

File: web/src/model/order.js
import {addMinutes} from '../util/date'
import {Product} from './product'

const orderTimeoutMin = 10

class Order {
  constructor({id, queueId, createdAt, updatedAt, status, eta, totalSum, products}) {
    this.id = id
    this.queueId = queueId
    this.createdAt = new Date(createdAt)
    this.updatedAt = new Date(updatedAt)
    this.status = status
    this.eta = new Date(eta)
    this.totalSum = totalSum

    if (products) this.products = products.map(Product.new)
    else this.products = []
  }

  get isPending() {
    return this.status === 'pending'
  }

  get isDelivering() {
    return this.status === 'delivering'
  }

  get isFulfilled() {
    return this.status === 'fulfilled'
  }

  get isRejected() {
    return this.status === 'rejected'
  }

  get isTimedOut() {
    return addMinutes(new Date(), -orderTimeoutMin) >= this.createdAt
  }

  static new(data) {
    return new Order(data)
  }
}

export {
  Order
}
